Fault logs should not be skipped

pvollan opened the pull request:
#### 4e7db72c941ca57450ff0a3366f1e756b47235c4
<pre>
Fault logs should not be skipped
<a href="https://bugs.webkit.org/show_bug.cgi?id=258577">https://bugs.webkit.org/show_bug.cgi?id=258577</a>
rdar://111393438

Reviewed by Chris Dumez.

Fault logs should not be skipped in the WebContent process, but sent asap to the Networking process.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogHook):

Canonical link: <a href="https://commits.webkit.org/265558@main">https://commits.webkit.org/265558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ddaee3fb6e95c1318d9ae17174d9804170826fed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11230 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11440 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11773 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12881 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10710 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11425 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13624 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11391 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12276 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9490 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13290 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9570 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17366 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10640 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10321 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13542 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10751 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8836 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9921 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2690 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14195 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10604 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -723,14 +723,30 @@ static void registerLogHook()
             return;
 
-        // Skip faults and debug logs in non-internal builds.
-        if (type == OS_LOG_TYPE_FAULT || type == OS_LOG_TYPE_DEBUG)
```
Oh wow, this is an important fix lol

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -723,14 +723,30 @@ static void registerLogHook()
+
+            auto connectionID = WebProcess::singleton().networkProcessConnectionID();
+            if (connectionID)
+                IPC::Connection::send(connectionID, Messages::NetworkConnectionToWebProcess::LogOnBehalfOfWebContent(logChannel.bytesInludingNullTerminator(), logCategory.bytesInludingNullTerminator(), logString, OS_LOG_TYPE_ERROR, getpid()), 0);
```
I suggest with still use the dispatch queue but use WorkQueue::dispatchWithQoS() to make the message priority higher. Otherwise, we can end up with out-of-order logging which would be unfortunate.

cdumez replied:
We should also pass the QoS to IPC::Connection::send() for FAULT messages (already supported).

pvollan replied:
Ah, good point, will fix!

Thanks for reviewing!

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -723,20 +723,28 @@ static void registerLogHook()
+
+        // Send fault logs with high priority. If the WebContent process is terminated, we might not be able to send the log in time.
+        if (type == OS_LOG_TYPE_FAULT) {
+            type = type == OS_LOG_TYPE_ERROR;
```
What's this?

pvollan replied:
That was a typo, fixed in latest patch :)

cdumez reviewed (approved):
LGTM

pvollan committed:
Fault logs should not be skipped
https://bugs.webkit.org/show_bug.cgi?id=258577
rdar://111393438

Reviewed by Chris Dumez.

Fault logs should not be skipped in the WebContent process, but sent asap to the Networking process.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogHook):

Canonical link: https://commits.webkit.org/265558@main
